### PR TITLE
Move to use JavaFX

### LINKED ISSUE #2: Make boolean values a checkbox
It's my primary complaint about OutlineViewer.

### PR DESCRIPTION
**Pull request by SamCarlberg**
This also fixes #2, #6, #14, #19, #20, #21, and #27, supersedes #7 and #22, and adds the following features:

- Change app settings on the fly
  - Show/hide metadata
  - Change server address
  - Change server port
  - Change from server mode to client mode, and back again
- Search for keys/values/types
- Delete keys and branches with the delete key (no menu needed)
- Multiple selection of keys for easy deletion

Requires wpilibsuite/ntcore#202. Jenkins builds will fail until that PR is merged and the network tables dependencies in `build.gradle` are reverted to use the desktop jar instead of the one I'm building locally.

Currently this is a mess (see: MainWindowController) and needs a lot of cleanup (and tests, and checkstyle, and...)

Still TODO:

- [x] Support raw data
- [x] Support string arrays
- [x] Support number arrays
- [x] Support boolean arrays
- [ ] Comprehensive tests

### Preferences (server mode)
screenshot from 2017-05-29 21 35 20

### Preferences (client mode)
screenshot from 2017-05-29 21 35 30

### Connecting to a remote server
screenshot from 2017-05-29 21 39 18

### No server connection
screenshot from 2017-05-29 21 39 43

### Running a server locally
screenshot from 2017-05-29 21 40 59

### Add or edit strings and numbers
screenshot from 2017-05-29 21 41 07

### Add or edit booleans
screenshot from 2017-05-29 21 41 21

### String / number arrays
screenshot from 2017-05-29 21 41 43

### Boolean arrays
screenshot from 2017-05-29 21 42 12

### Arrays of raw bytes
screenshot from 2017-05-29 21 42 54

### CODE REVIEW

**Review by AustinShalit — APPROVED** [review]

**Comment by ThadHouse** [comment]
Where does setting the remote name happen? Also that only works before a connection starts.

**Comment by SamCarlberg** [comment]
There's a bit going on under the hood. If the server ID is a number, it assumes it's a team number and attempts to connect to `roborio-TEAM-frc.local`. Otherwise it's assumed to be a URL or IP address and passes that directly to ntcore.

Changing modes between client and server shuts down the current instance of ntcore and spins up a new one. Same for changing port or server ID.

**Comment by ThadHouse** [comment]
I meant remote name. Since you tagged #14. That option should be added to that window as well.

**Comment by ThadHouse** [comment]
Also if possible we should support arrays of IP addresses for connecting to, since thats supported by ntcore as well.

File: src/main/java/edu/wpi/first/tableviewer/PreferencesController.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+      port = Integer.parseInt(portField.getText());
+    }
+    if (url.isEmpty()) {
+      url = "localhost";
+    }
+    url = url.replaceAll("^.*://", ""); // remove leading protocol
+    if (url.matches("[0-9]+")) {
+      // treat as a team number
+      url = "roborio-" + url + "-frc.local";
+    }
```
Comment (AustinShalit) [inline src/main/java/edu/wpi/first/tableviewer/PreferencesController.java]: Use nt set team here instead. That way it will try all the possible ips

Comment (SamCarlberg) [reply src/main/java/edu/wpi/first/tableviewer/PreferencesController.java]: I forgot ntcore had that. Thanks!

Comment (AustinShalit) [reply src/main/java/edu/wpi/first/tableviewer/PreferencesController.java]: Also, you can change the regex to limit the amount of numbers to be less than 4

Comment (SamCarlberg) [reply src/main/java/edu/wpi/first/tableviewer/PreferencesController.java]: I think `"[0-9]{1,5}"` would be fine. 0-5 digits is enough for a while

**Comment by AustinShalit** [comment]
@thadhouse Are you thinking a comma separated list for IP addresses?

**Comment by ThadHouse** [comment]
@AustinShalit sure they could work. I know we were discussing parallelizing connections in ntcore, so adding more isn't an issue.

**Comment by SamCarlberg** [comment]
I feel like multiple connections is out of scope for this project

**Comment by ThadHouse** [comment]
True. We can add that at a later date. Along with remote name if thats too much.

**Comment by SamCarlberg** [comment]
@JLLeitschuh Any idea why the travis UI tests are failing? Looks like there's no JavaFX installed in the image

**Comment by AustinShalit** [comment]
Try this: https://github.com/wpilibsuite/OutlineViewer/pull/24/files#diff-354f30a63fb0907d4ad57269548329e3

File: src/main/java/edu/wpi/first/tableviewer/Main.java [hunk 2]
```diff
@@ -39,6 +39,8 @@ public void start(Stage primaryStage) throws IOException {
     primaryStage.setOnCloseRequest(e -> System.exit(0));
     primaryStage.setScene(new Scene(mainWindow));
     primaryStage.show();
+    primaryStage.setMinWidth(600);
+    primaryStage.setMinHeight(450);
```
Comment (AustinShalit) [inline src/main/java/edu/wpi/first/tableviewer/Main.java]: Should this be in FXML?

**Pull request merged**